Bookings: Update selector for service/event and customer name filters

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses. -->
<!-- Part of: # -> use this if your PR is one of many related to one issue -->

Part of WOOMOB-1240

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR updates the selector for Service/Event and Customer name filters for bookings:
- Adds a new type conforming to `ListSyncable` to support showing selector for bookable products (named as Service/Event on the UI).
- Updates the Customer selector:
  - Renames the title to match the filter option for bookings
  - Updates the navigation to use push instead of modal.
  - Adds new configuration for the presentation type to ensure that the selector is still presented modally from order creation form.
  - Update UI for the customer rows to better match the design. I'm keeping the search bar for simplicity.
  - Avoid dismiss selector upon selection. We'll need to support multiple selection in a subsequent PR.

## Testing steps
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

- Log in to a CIAB site with existing bookings.
- Navigate to Bookings > All > Filter.
- Select Service/Event and confirm that only bookable products are displayed. Selecting any option should show a checkmark next to the option. Dismissing the view should display the product name on the filter screen.
- Select Customer name and confirm that the customer selector is pushed from the same navigation stack. 
- Confirm that the customer selector has the correct title and selecting any option would show a checkmark next to the option. Dismissing the view should display the customer name on the filter screen.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- Tested and confirmed with simulator iPhone 17.
- Also checked the customer selector on Order filter and Order creation form and confirmed that no regression was made.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->



https://github.com/user-attachments/assets/fbc9a8ef-2e33-4463-88ae-8bfb59ae70cd




---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Add new view for selecting service/event

itsmeichigo committed:
Update customer filter

itsmeichigo committed:
Move Customer name row down

itsmeichigo committed:
Update layout for customer selector rows

itsmeichigo committed:
Hide username for customers in booking filter

itsmeichigo committed:
Do not dismiss selector views after selection

adborbas commented:
Hi @itsmeichigo, There might be something I'm missing, but I could not get the selections dismiss working.

| Case | Recording | Description |
| --- | --- | --- |
| Service / Event selection | ![Simulator Screen Recording - iPhone 17 Pro - 2025-10-27 at 09 14 34](https://github.com/user-attachments/assets/8a9b9dde-5423-49a8-b2a9-c69ffa02368d) | Selection does not dismiss the view |
| Customer selection | ![Simulator Screen Recording - iPhone 17 Pro - 2025-10-27 at 09 15 22](https://github.com/user-attachments/assets/43d52ddf-e98d-42d6-9709-af0e23fc2fbf) | Selection does not dismiss the view |
| No name customer selection | ![Simulator Screen Recording - iPhone 17 Pro - 2025-10-27 at 09 15 37](https://github.com/user-attachments/assets/1b314ed9-5045-4af6-bcfc-3acab315d912) | Selecting a "No name" customer after a "No name" customer changes the order. <li/> Also, there is no option to clear selection |

itsmeichigo committed:
Update sort descriptors for customers

itsmeichigo commented:
Thanks @adborbas for the feedback!

> I could not get the selections dismiss working.

I just updated all the filters for bookings in this PR to not dismiss the selector immediately upon selection. This will keep the behavior consistent with the static options as well (attendance status, booking status). Also we'll need to support multiple selections, but that will be handled later. So for now, one would need to either tap < arrow to add more filter or tap Show bookings to apply the filters.

> Selecting a "No name" customer after a "No name" customer changes the order.

I updated the sort descriptor in fb766876a48a06e4a829fd6c5bf1939fa46dc200 to fix this.

> there is no option to clear selection

I asked Wagner in lxixW6aBBNkCd9ooBK70Z1-fi-1421_3698#1478520395 - I'll update this later when I get a response from him.

adborbas commented:
> I just updated all the filters for bookings in this PR to not dismiss the selector immediately upon selection.

@itsmeichigo Ok, thanks for the clarification! This makes sense now. 

In this case I think the "Testing steps" needs some adjustment cause it says:
- "Selecting any option should dismiss the view and display the product name on the filter screen."
- "Selecting any option would dismiss the view and display the customer name on the filter screen."

I'd expect them to say _not_ dismiss the view.

I'll check the code now.

adborbas commented on Modules/Sources/Yosemite/Model/Bookings/BookingProductFilter.swift:
```diff
@@ -1,20 +1,23 @@
 ///
 public struct BookingProductFilter: Codable, Hashable {
+    /// The underlying product
+    public let product: Product
```
One small concern — I wonder if we really need to store the entire Product here. Having both product, id, and name feels a bit redundant and might make it easier for other parts of the code to start depending on BookingProductFilter for full product data instead of just treating it as a lightweight filter value.

Would it make sense for this type to only keep the minimal info needed for filtering (e.g. id and name) and drop the product property? That would keep the model focused on its purpose and avoid any unintended coupling later on.

itsmeichigo replied:
This is a valid point! I added an update in b6ed021b9b629cdd351881b7194af15b61bdf96a with the following changes:
- Removed the full product model from `BookingProductFilter`.
- Updated `SyncableListSelectorView` to accept a closure for checking initial selection rather than comparing full models.
- Updated `FilterListViewController` following the selection logic change.

adborbas reviewed (approved):
Nice work on this 👍 Had one small comment.

itsmeichigo committed:
Update initial selection logic and remove full product model from BookingProductFilter

itsmeichigo commented:
> I'd expect them to say not dismiss the view.

My bad, I updated the code but forgot to update the PR description. I updated it to avoid confusion in case someone else visits the PR.

I just addressed the issue you found above, will enable automerge now to unblock my other PR. Please feel free to leave comments if any, I'll address them in subsequent PRs.

Thanks!
